zebrad/src/components/inbound.rs

```diff
@@ -145,64 +152,58 @@ impl Service<zn::Request> for Inbound {
-                            network_setup,
-                            verifier,
-                        };
-                    }
-                    Err(error @ TryRecvError::Closed) => {
-                        // Mark the service as failed, because network setup failed
-                        error!(?error, "inbound network setup failed");
-                        let error: SharedRecvError = error.into();
-                        self.network = Setup::FailedRecv {
-                            error: error.clone(),
-                        };
-                        return Poll::Ready(Err(error.into()));
+                }
+                Err(TryRecvError::Empty) => {
+                    // There's no setup data yet, so keep waiting for it
+                    Setup::AwaitingNetwork {
+                        network_setup,
+                        verifier,
                     }
                 }
+                Err(error @ TryRecvError::Closed) => {
+                    // Mark the service as failed, because network setup failed
+                    error!(?error, "inbound network setup failed");
+                    let error: SharedRecvError = error.into();
+                    result = Err(error.clone().into());
+                    Setup::FailedRecv { error }
+                }
+            },
+            // Make sure we left the network setup in a valid state
+            Setup::FailedInit => unreachable!("incomplete Inbound initialization"),
+            // If network setup failed, report service failure
+            Setup::FailedRecv { error } => {
+                result = Err(error.clone().into());
+                Setup::FailedRecv { error }
             }
-        }
-
-        // Unfortunately, we can't combine these matches into an exhaustive match statement,
-        // because they use mutable references, or they depend on the state we've just modified.
-
-        // Make sure we left the network setup in a valid state
-        if matches!(self.network, Setup::FailedInit) {
-            unreachable!("incomplete Inbound initialization");
-        }
-
-        // If network setup failed, report service failure
-        if let Setup::FailedRecv { error } = &mut self.network {
-            return Poll::Ready(Err(error.clone().into()));
-        }
-
-        // Clean up completed download tasks, ignoring their results
-        if let Setup::Initialized { downloads, .. } = &mut self.network {
-            while let Poll::Ready(Some(_)) = downloads.as_mut().poll_next(cx) {}
-        }
+            // Clean up completed download tasks, ignoring their results
+            Setup::Initialized {
+                address_book,
+                mut downloads,
+            } => {
+                while let Poll::Ready(Some(_)) = downloads.as_mut().poll_next(cx) {}
+                Setup::Initialized {
+                    address_book,
+                    downloads,
+                }
+            }
+        };
 
```

Comment:
We should panic if `self.network` is `FailedInit` after the match and assignment, because that's an invalid state.

Reply:
Fixed in 9bc8170c27376421d41c5633fdd4bd99ed8a538e